Automattic/pocket-casts-android
# Track global episode events - Part 1

## Description

- This tracks below episode events:
    - `episode_download_queued`: When the user triggers episode for downloading
    - `episode_download_finished`: When episode download is finished
    - `episode_bulk_download_queued`: When the user triggers multiple episodes download while in multi select mode
    - `episode_download_cancelled`: When the user cancels an active download
    - `episode_upload_queued`: When the user uploads a custom file to the cloud
    - `episode_upload_cancelled`: When the user cancels an active cloud upload
    - `episode_upload_finished`: When episode upload is finished
    - `episode_deleted_from_cloud`: When the user deletes a file from the cloud
    - `episode_download_deleted`: When the user deletes a downloaded episode
    - `episode_bulk_download_deleted`: When the user deletes multiple downloaded episodes while in multi select mode
- Since episode sources are spread across multiple screens and are mostly common to playback sources, `PlaybackSource` is extracted to a common `AnalyticsSource` and reused as episode source as well - 73af404133a8e136f6c8548041f62baa1d23d080.
- Other events covered in corresponding [iOS PR](https://github.com/Automattic/pocket-casts-ios/pull/298) will be added as part of another PR.

------




## Testing Instructions
<details><summary>Download / Bulk Download / Download Cancel</summary>

1. Tap on the Podcasts tab
2. Tap on a podcast
3. Tap on an episode
4. Tap the download button
5. ✅ `🔵 Tracked: episode_download_queued ["source": "episode_detail", "episode_uuid": "UUID"]`
7. Wait for it to finish downloading, 
8. ✅ `🔵 Tracked: episode_download_finished ["source": "episode_detail", "episode_uuid": "UUID"]`
9. Tap the download button again to delete it
10. ✅ `🔵 Tracked: episode_download_deleted ["episode_uuid": "UUID", "source": "episode_detail"]`
11. Tap the download button again, and then tap it again to cancel the download before it finishes
12. ✅ `🔵 Tracked: episode_download_cancelled ["episode_uuid": "UUID", "source": "episode_detail"]`
13. Dismiss the episode details
14. Tap and hold on the podcast list to enter multi select mode
15. Select multiple episodes
16. Tap the ... if the Download action is not in your shelf
17. Tap the Download action
18. ✅ `🔵 Tracked: episode_bulk_downloaded ["episode_count": COUNT, "source": "podcast_screen"]` - Verify the count is the number of episodes you selected
19. Enter multi select mode again
20. Select the same episodes
21. Tap the Remove Download action
22. ✅ `🔵 Tracked: episode_bulk_download_deleted ["episode_count": COUNT, "source": "podcast_screen"]`

**Repeat the steps on the following views:**
1. Filter Details
2. Downloads
3. Starred
4. Listening History
5. Files
</details>

<details><summary>Episode Upload / Cancelled / Deleted</summary>

1. Tap on the Profile icon
2. Tap on Files
3. If you don't have any custom files, add a couple by dragging and dropping an audio file from your Mac to the simulator
4. Tap on a file in the list
5. Tap Upload to Cloud
6. ✅ `🔵 Tracked: episode_upload_queued ["episode_uuid": "UUID", "source": "files"]`
7. Wait for it to finish uploading, 
8. ✅ `🔵 Tracked: episode_upload_finished ["episode_uuid": "UUID", "source": "files"]`
8. Tap the same file again
9. Tap the Remove from Cloud action
10. ✅ `🔵 Tracked: episode_deleted_from_cloud ["episode_uuid": "UUID", "source": "files"]`
11. Tap the file again, and select upload to cloud
12. While the file is uploading, tap it again and select the Cancel Download action
13. ✅ `🔵 Tracked: episode_upload_cancelled ["episode_uuid": "UUID", "source": "files"]`
</details>
